torchtune/modules/attention.py

```diff
@@ -10,6 +10,85 @@
 import torch
 from torch import nn, Tensor
 from torchtune.modules.kv_cache import KVCache
+from torchtune.utils._version import torch_version_ge
+from torchtune.utils.attention_bias import _MaskType
+from torchtune.utils.logging import get_logger, log_once
+
+_log: logging.Logger = get_logger()
+
+
+if torch_version_ge("2.5.0"):
+    from torch.nn.attention.flex_attention import BlockMask, flex_attention
+
+    flex_attention_compiled = torch.compile(flex_attention, dynamic=False)
```

Comment:
Not a review yet, just a reminder that this should probably be controlled by cfg.compile. If compile=False, and we use flex_attention, maybe it  should raise an error, or default back to efficient attention

Reply:
@kartikayk mentioned that there are machines that are not compatible with compile. 

option 1: raise and error

options 2: let users control compile_model, compile_loss, compile_attention (i dont like it too much)

option 3:This may be too much, but its probably the optimal way. Today we work with SDPA, and if the mask is not causal, we default back to efficient attention, right?

We could add flexattention to this logic.

if mask == is_causal:
    att = SPDA
elif compile==True:
     att = flexattention
else:
     att = efficient_attention

Reply:
I don't think any of our components should have a hard dependency on compile since this limits usability to backends which don't support compile. I think any design should account for this and provide escape hatches. I prefer the flash-attention approach of falling back to the less performant option rather than erroring out. In this case the "less performant" way would actually be to do what we do today i.e. no FlexAttention instead of flex attention without compile (unless you have evidence that flex without compile is ok). Is there a way we can support multiple attention APIs - kinda like option 3 from @felipemello1

Reply:
I see - we can do a variant of option 3 that also checks for pytorch version. it will be messy, but I don't see any other way

Reply:
I do not reccomend using FlexAttention without compile, the EfficientAttention backend and soon CuDNN backend with full bias support will have the same memory usage and be much more performant as using flex in eager

Reply:
So the scenarios where compilation will fail are pretty much the same for triton - older gpus. Efficient attention has slightly better older gpus suppport. I would say if you are newer than v100 flex should work and v100 or older (in terms of sm version) efficient would work

Reply:
ok, update from Driss:

> [https://github.com/pytorch/pytorch/blob/758d78790164bfb041555daed380de96e06f78a3/torch/testing/_internal/common_cuda.py#L29-L32](https://l.workplace.com/l.php?u=https%3A%2F%2Fgithub.com%2Fpytorch%2Fpytorch%2Fblob%2F758d78790164bfb041555daed380de96e06f78a3%2Ftorch%2Ftesting%2F_internal%2Fcommon_cuda.py%23L29-L32&h=AT2SiPER_AEu74JpqaJrnqKSKKj3VJRl_FKR0wiKuL_fYyOrMlaDwNocwYxfQ7Rh_4YlepFVd4JKK5WfOVrlUf-kpJv1eSFDrasGq7E-LG5nu7DvsAkLQntV2xFrhsEsKsRyYZqyOddN0pWqirHHEz0uFOtIuh3sIq3mog)
> 
> ignore the lazyval stuff and just do:
> 
> should_use =  torch.cuda.is_available() and torch.cuda.get_device_capability() >= (7, 5)
> + [https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/](https://l.workplace.com/l.php?u=https%3A%2F%2Farnon.dk%2Fmatching-sm-architectures-arch-and-gencode-for-various-nvidia-cards%2F&h=AT2SiPER_AEu74JpqaJrnqKSKKj3VJRl_FKR0wiKuL_fYyOrMlaDwNocwYxfQ7Rh_4YlepFVd4JKK5WfOVrlUf-kpJv1eSFDrasGq7E-LG5nu7DvsAkLQntV2xFrhsEsKsRyYZqyOddN0pWqirHHEz0uFOtIuh3sIq3mog)
> 
> Honestly might be a question for: https://fb.workplace.com/groups/385893200869952
> 
> As to what they recommend the min version that should be used

Reply:
ok, to summarize the discussion here, only run flex attention if the following are true:
- We are sample packing
- User has torch >= 2.5
- torch.cuda.get_device_capability() >= (7, 5)

and since it doesn't make sense to run flex without compile, we will just compile by default since we are already checking for a certain GPU capability

Reply:
Right and if we are sample packing and either of (2) or (3) don't hold then we will just fall back to SDPA with memory-efficient attention